include/http/router.js

```diff
@@ -1,197 +1,276 @@
-const Router = require('koa-router');
-const Session = require('../koa/Session')();
-const bodyParser = require('koa-body');
-const Cookies = require('koa-cookie').default;
-const Passport = require('../koa/authentication/Passport')();
-const https = require('https');
-const fs = require('fs');
+'use strict';
 
-module.exports = function(pb) {
+//dependencies
+var url = require('url');
+var util = require('util');
+var domain = require('domain');
 
-    class PencilblueRouter {
+module.exports = function (pb) {
 
-        constructor() {
-            this.app = new Koa();
-            this.app.keys = ['9011fa34-41a6-4a4d-8ad7-d591c5d3ca01']; // Random GUID
-            this.app._requestsServed = this.app._requestsServed || 0;
+    //pb dependencies
+    var RequestHandler = pb.RequestHandler;
 
-            this.router = new Router();
+    /**
+     * Responsible for routing a request through the registered middleware to serve up a response
+     * @class Router
+     * @constructor
+     * @param {Request} req
+     * @param {Response} res
+     */
+    class Router {
+        constructor(req, res) {
 
-            this.app.use(bodyParser({
-                multipart: true,
-                // formidable: { uploadDir: path.join(__dirname, 'tmp') }
-            }));
-            let passport = Passport(pb);
-            this.app.use(Session(this.app));
-            this.app.use(Cookies());
-            this.app.use(passport.initialize());
-            this.app.use(passport.session());
+            /**
+             * Represents the current position of the middleware that is currently executing
+             * @property index
+             * @type {number}
+             */
+            this.index = 0;
+
+            /**
+             * @property req
+             * @type {Request}
+             */
+            this.req = req;
+
+            /**
+             * @property res
+             * @type {Response}
+             */
+            this.res = res;
         }
 
-        static registerRoute(routeDescriptor) {
-            this.internalRouteList.push(routeDescriptor);
+        /**
+         * Starts the execution of the middleware pipeline against the specified request/response pair
+         * @method handle
+         */
+        handle() {
+
+            //set reference to the handler
+            this.req.handler = new RequestHandler(this.req, this.res);
+            this.req.router = this;
+
+            return this._handle(this.req, this.res).catch((err) => {
+                this.res.statusCode = 500;
+                this.res.end();
+            });
         }
 
-        static addMiddlewareAfter(name, middleware) {
-            let index = this._indexOfMiddleware(name);
-            if (index >= 0) {
-                return this._addMiddlewareAt(index + 1, middleware);
+        _handleMiddleware(req, res, action) {
+            return new Promise((resolve, reject) => {
+                process.nextTick(() => {
```

Comment:
is process.nexTick really necessary here? since with the await in the for where we are callling the function will ensure that  the action execution is finished before execute the next middleware

Reply:
I want to say the original domain was in a process.next tick.  Pencilblue used it all over in their original callback code when handling server requests.